chain/penumbra/penumbra_app_node.go

```diff
@@ -0,0 +1,219 @@
+package penumbra
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"runtime"
+	"strings"
+	"time"
+
+	"github.com/ory/dockertest"
+	"github.com/ory/dockertest/docker"
+	"github.com/strangelove-ventures/ibc-test-framework/ibc"
+	"github.com/strangelove-ventures/ibc-test-framework/utils"
+)
+
+type PenumbraAppNode struct {
+	Index     int
+	Home      string
+	Chain     ibc.Chain
+	TestName  string
+	NetworkID string
+	Pool      *dockertest.Pool
+	Container *docker.Container
+}
+
+const valKey = "validator"
+const tendermintPort = "26658/tcp"
+
+var exposedPorts = map[docker.Port]struct{}{
+	docker.Port(tendermintPort): {},
+}
+
+// Name is the hostname of the test node container
+func (p *PenumbraAppNode) Name() string {
+	return fmt.Sprintf("node-%d-%s-%s", p.Index, p.Chain.Config().ChainID, p.TestName)
+}
+
+// Dir is the directory where the test node files are stored
+func (p *PenumbraAppNode) Dir() string {
+	return fmt.Sprintf("%s/%s/", p.Home, p.Name())
+}
+
+// MkDir creates the directory for the testnode
+func (p *PenumbraAppNode) MkDir() {
+	if err := os.MkdirAll(p.Dir(), 0755); err != nil {
+		panic(err)
+	}
+}
+
+// Bind returns the home folder bind point for running the node
+func (p *PenumbraAppNode) Bind() []string {
+	return []string{fmt.Sprintf("%s:%s", p.Dir(), p.NodeHome())}
+}
+
+func (p *PenumbraAppNode) NodeHome() string {
+	return fmt.Sprintf("/tmp/.%s", p.Chain.Config().Name)
+}
+
+func (p *PenumbraAppNode) CreateKey(ctx context.Context, keyName string) error {
+	cmd := []string{"pcli", "wallet", "generate"}
+	exitCode, stdout, stderr, err := p.NodeJob(ctx, cmd)
+	// already exists error is okay
+	if err != nil && !strings.Contains(stderr, "already exists, refusing to overwrite it") {
+		return utils.HandleNodeJobError(exitCode, stdout, stderr, err)
+	}
+	cmd = []string{"pcli", "addr", "new", keyName}
+	return utils.HandleNodeJobError(p.NodeJob(ctx, cmd))
+}
+
+func (p *PenumbraAppNode) GetAddress(ctx context.Context, keyName string) ([]byte, error) {
+	cmd := []string{"pcli", "addr", "list"}
+	exitCode, stdout, stderr, err := p.NodeJob(ctx, cmd)
+	if err != nil {
+		return []byte{}, utils.HandleNodeJobError(exitCode, stdout, stderr, err)
+	}
+	addresses := strings.Split(stdout, "\n")
+	for _, address := range addresses {
+		fields := strings.Fields(address)
+		if len(fields) < 3 {
+			continue
+		}
+		if fields[1] == keyName {
+			// TODO looks like penumbra address is bech64? need to decode to bytes here
```

Comment:
Penumbra addresses use Bech32m encoding: https://github.com/penumbra-zone/penumbra/blob/main/crypto/src/address.rs#L156 (a bit more detail here on the encoding in case it's useful: https://protocol.penumbra.zone/main/protocol/addresses_keys/addresses.html#address-encodings)

Reply:
thanks! looks like I don't need to decode it yet actually to get the chain moving, but updated the TODO message for later when it's needed for IBC testing.